### PR TITLE
Simplifications and improvements to DeepZipper and merge strategies

### PR DESCRIPTION
**Pull request by josharnold52**
This patch makes some simplifications to the `DeepZipper` implementation.  Documentation has also been for the `unselect` algorithm and the notion of "merge strategies" has been fleshed out.

Some highlights:
1.  The new [DeepZipper](https://github.com/josharnold52/anti-xml/blob/zippermods/src/main/scala/com/codecommit/antixml/DeepZipper.scala) trait is significantly shorter.  I think it is easier to follow as well. 
2.  The unselection algorithm is defined in the `DeepZipper` scaladoc.
3.  Merge strategies are now traits and can be supplied to `unselect` via an implicit parameter.  Some additional predefined 
   strategies [have been provided](https://github.com/josharnold52/anti-xml/blob/zippermods/src/main/scala/com/codecommit/antixml/ZipperMergeStrategy.scala).
4.  The alternatives presented to a Merge strategy are explicitly defined.
5.  The handling of node duplication and ellision is more uniform.  This accounted for much of the simplification and also
   addressed some potentially surprisng behavior in the unselect algorithm (I'll log a separate issue for that).

With one exception, no functionality was lost.  In fact, these changes enable some merge strategies that weren't possible before.  

The exception is that the merge strategy is now _only_ called at "conflict" locations where both a node and its descendant have been selected.   In the old algorithm, the merge strategy was also called at non-conflicting locations, but the alternatives sequence would only contain a single entry.

The change was matter of preference.  I can't think of many realistic situations that would require the strategy to be called at nodes where there are no conflicts.  But I can easily switch the algorithm back to the old behavior if that is desired.   The `ZipperMergeStrategy` API would need to be changed slightly, but the implementation change would be trivial.

### CODE REVIEW

**Comment by djspiewak** [comment]
Merged; thanks!

**Comment by ncreep** [comment]
Wow! I was suspecting that `DeepZipper` can/should be simplified, but I didn't expect it to become so terse...
Making it more mathematical like, really makes a difference.

I'm still trying to figure out how it works, but I got one question though:
I saw that you removed the limitation that paths cannot contain duplicate locations. How exactly does the new implementation handles the cases where this limitation is broken?

**Comment by josharnold52** [comment]
I'm not sure I understand the question, possibly because we are attaching slightly different meanings to the terms "path" and "location".     By "path", I just mean a sequence of Ints that takes you from the top-level group down to some descendant.   I'll just avoid using "location" for now.

My mental model for a Zipper is that it defines a "pullback" function from paths to a sequence of `Nodes`,  which are the replacements for the original node at that path in the XML tree.    Given that, `unselect` is basically a process of replacing each node with the sequence of nodes given by its pullback.

The important point is that we always consider the pullback to yield a _sequence_.   That eliminates most problems with duplicates because we can always just concatenate the corresponding sequences together.

For example, if the Zipper builder is given path P with nodes (N1, N2, and N3) and then later is given the same path P with nodes (N11 and N12), it merges them such that path P is now associated with (N1, N2, N3, N11, and N12).   This happens in the `Zipper.WithZipperBuilder` class as it builds the `pathIndex` Map.

Does that answer your question?

**Comment by ncreep** [comment]
OK, that does answer my question, but I think that this aspect of the implementation is slightly problematic.

I'm not sure that concatenation is necessarily the right thing to do here.
Consider a `Group` containing a single element and a selection function that somehow yields that very same element twice.
Unselecting on the resulting `Zipper` would yield a new `Group` with two identical elements. 
Although I'm not sure what would a user expect to get in this situation, it should at least be pluggable, i.e. the resolution of this conflict should be delegated to the merging strategy.

The reason why in the first place I decided not to allow duplicate paths (or "locations" in my terminology) is that I couldn't think of a nice way to resolve conflicts between two or more sequences that are attached to the same path (which would be obtained by flatMapping).

So I think it would be safer to just forbid the creation of `Zippers` with more than a single sequence attached to a path.
Or, which would make the whole thing more general, and will probably ease the creation of more intricate selection functions, to find a way to propagate the different sequences down to the merging strategy.

**Comment by josharnold52** [comment]
I don't see this as a problem.  Of course, none of our selection operators (`\`, `\\`, `\\!`, and `select`) ever produce duplicate paths so maybe I'm just having trouble envisioning how this situation would ever turn up.  Do you have an example in mind?

The current code actually does forbid the creation of Zippers with more than one sequence attached to a path, because there's no way to ever specify one!   CanBuildFromWithZipper simply allows you to build a path's sequence in chunks.   That was by design and Zipper.flatMap takes advantage of this behavior.

Of course, if there is ever a need to associate multiple sequences for a path, then CanBuildFromWithZipper or something like it would have to be enhanced to allow such a thing to be specified.  But I don't think we have to tackle that until we actually want to support that functionality.

**Comment by ncreep** [comment]
I don't really have anything concrete. 
I was thinking a little about selection of node siblings, and although I'm not sure, I think that sibling selection could quite easily yield duplicates. For example, selecting all the siblings of a `Group` of nodes, where some of the nodes in the `Group` are siblings of each other.

Anyways, in my view, if the API can be abused, it will be abused, and a run time error seems like a reasonable solution, at least until the issue is resolved.

**Comment by josharnold52** [comment]
I don't see what API there is to abuse.  The behavior of `CanBuildFromWithZipper` isn't a loophole.  It's intentional, [documented behavior](http://www.danielspiewak.com/anti-xml/v0.3/doc/#com.codecommit.antixml.CanBuildFromWithZipper$$ElemsWithContext) and the current `Zipper.flatMap` relies on it.    Maybe you feel the API is confusing.  There were specific reasons behind it, but I'm not entirely happy with the result and I'd love it if we can simplify it.    However, I think there will always be a need for `CanBuildFromWithZipper` to associate multiple `Nodes` with a path\location.

Correct me if I'm wrong, but didn't the old `CanBuildFromWithDeepZipper` allow duplicates as well?   I know there was a `fromPath` method that forbade duplicates, but that was `private[antixml]` and I think it was only used by the selection operators, which couldn't have produced duplicates anyway.

Of course, if you are envisioning a more general selection API, then maybe enforcing a no-duplicates policy will be appropriate there.  But again, I think we can cross that bridge when we come to it.

**Comment by ncreep** [comment]
The API abuse that I see is that we're exposing an implementation detail that users may come to rely upon.

You're quite right that the old `CanBuildFromWithDeepZipper` was allowing this kind of behavior, but that was only to satisfy `flatMap` and I couldn't find a way around it.
In the case of the old implementation, the main method to construct `Zippers` was to use a path function, as using `CanBuildFromWithDeepZipper` was not as straightforward (not that it is a valid excuse), so placing the limitation there seemed like a reasonable compromise.

In the end, I think that throwing an exception when constructing the paths, is a good reminder to developers/users about this limitation of the `Zipper`, which in my eyes is not that evident from the documentation.

**Comment by josharnold52** [comment]
But there _was_ no public API for path functions, nor is there now.  `DeepZipper.fromPathFunc` was `private[antixml]`, as was`PathCreator`.    Users couldn't have been using it without adding to the `antixml` package, which they do at their own risk.    Would it address your concern if we added an `assume` check for uniqueness in `Selectable.fromPathFunc`?  That would remind us to revisit this if we ever open these methods up. 

From a more general perspective, I wonder if we have different visions for what the "ideal" Zipper should do (which would be great because that makes it interesting!)   I see the Zipper as an association between a subset of paths (a.k.a locations, a.k.a holes) in the original tree and a _single_ subset of items in the Zipper sequence.  Upon unselect, each subset gets inserted into its hole in the Zipper's order.  I don't see this as a limitation or implementation detail; I see it as the most natural definition that closes over selection and `flatMap`. It also yields a completely natural `unselect` provided there is no lineal relationship between paths.  Merge strategies are only required to deal with lineal conflicts.

My impression is that you would ideally have the Zipper associate a path/location/hole to _multiple_ subsets and have the merge strategy resolve them into a single sequence.   I can think of ways this could be done fairly cleanly, but I don't see why someone would want this in the first place.  In your example of sibling selection, I would rather the selection algorithm just prevent duplicates from occurring.  

In any case, I agree the documentation needs work!  I'm embarrassed to say I've already found mistakes in some of my scaladoc comments.  I'll work on some cleanup tonight.

**Comment by ncreep** [comment]
OK, maybe I slightly abused the term "API". I'm still a bit blurry about pretty much everything ranging from Scala to proper API design.
Basically, I'm a not so experienced Java/PHP developer, so coming to the world of Scala API design is a bit overwhelming.
The reason for the `private[antixml]` on `DeepZipper.fromPathFunc` does not stem from deep thoughts about APIs, but more from me copying what looks the most natural from the context around me.

Back to the point: yes, it would address my concern to use a uniqueness check on `Selectable.fromPathFunc`, and I would even say that the more appropriate check is `require` (but I wouldn't press on that). I do see us opening these methods up soon enough, e.g. trying to address XPath like axes and what not.

And now for the more general `Zipper` idea, my view is that a `Zipper` is the functional compromise for the fact that we can't use simple mutable trees. So the model for the perfect `Zipper` would be a one that would behave exactly like a mutable tree.

In my attempts to algorithmically define this "mutable tree behavior" I stumbled upon the problem of node identity, which is well defined in a mutable context, but not so much in an immutable one (I really like [this](http://clojure.org/state) idea, but I can't really see how to apply it to our case).
That's where the notions of "paths" and "time" come from, they are just tools to identify a node and what the user intended to do with it had he/she used a mutable tree. And the tool that actually performs this judgment is the merge strategy.
Now we could go for a more intricate notion of node "history" and store whole lists of node updates (a la mini Git like repository based on a tree) but I think that the added complexity wouldn't buy us much.

The reason why I think that duplicate paths will/should occur, is that given some arbitrary selection on a mutable tree, modifying one of the duplicate paths will auto-magically modify the rest of them. And that, in my opinion, is the most intuitive behavior. Trying to artificially limit selection, and restricting such scenarios, will always lead to some kind of counter-intuitive behavior, that will require justification (probably through massive documentation).
Although, this is probably unavoidable, we should strive to minimize these occasions, and keep the "perfect" mutable tree in mind - if a user of a mutable tree doesn't worry about duplicates, nor (ideally) should we.

(Having used the words "mutable tree" so many times makes me suspect that I might have an inferiority complex related to mutable tree users.)

**Comment by josharnold52** [comment]
Thanks for the clarifications & sorry for the delayed reply :-)

For now, I'll add some comments to `fromPathFunc` explaining the uniqueness issue for the benefit of any future public API.   As it stands, an actual check seems a bit wasteful given that it is clearly never called with duplicates.  I'll include the change along with some other documentation updates I'm planning, and we can revisit if you think it needs additional work.

I do agree that if we add a public API for selecting arbitrary paths, then it should prevent duplicates.  But there may be nicer ways to handle this than throwing an exception.  I think XPath automatically filters out duplicates and then orders the nodes lexically;  our API could do the same.  It could even accept a `Set` rather than a `Seq`, which would remove all ambiguity on the matter.  Of course, more general methods can always be added later if the need arises.

On the more philosophical front, I do view Zipper's purpose somewhat differently.  I don't think our aim should be to simulate multiply-referenced mutable state.  In fact, I think the Zipper is primarily useful in the following "singly-referenced" scenarios:
1.  When the Zipper context does not refer back to both a node and one its descendants.
2.  When the user of the Zipper only manipulates top-level properties of the Zipper's nodes and not their children.  

In these cases, there is a simple and intuitive way to pull updates back to the parent tree.  Things get hairier once you allow actual lineal conflicts.  The problem isn't that we're working with immutable data structures;  It's that we're trying to manipulate a tree structure using a sequence API.  The mutable analog has problems too.  Although some of the consistency issues go away, users have to contend with "spooky action at a distance" where updates to one index affect what is seen at another index.  This can easily be a source of bugs, and it's hard to account for because the sequence hides the ancestral relationships between the nodes.

If I had a good way to limit Zippers to just the simple cases, I'd probably advocate for it.  But as it stands, lineal conflicts seem to be a necessary evil.   Your idea of pluggable merge strategies is a good way to account for them, but in practice, I think sticking to the easy Zipper scenarios will result in simpler, more understandable code.

Beyond that, I don't see a need for allowing multiple references to the same location.  I don't know of a use case that requires them, and the generalization seems to complexify rather than simplify.   I think duplicates are problematic in the mutable case too.  IMO, a user of a mutable tree _does_ has to worry about duplicates! 

Fortunately, there's no pressing need to address the question of duplicates.  We can wait to see what people ask for :-)

**Comment by ncreep** [comment]
Using a sorted set is probably a good idea, though it may slightly hurt the generality of available selection methods.

Although I know that trying to imitate mutable data structures in a functional setting should count as pure evil, but that's where I take my intuitions from. Obviously intuitions change between different people, but I still think that when selecting on a tree, most people would (maybe naively) think of something along the lines of: "I want to select all elements with the name XXX and add the attribute YYY". When going through this process the thought of lineal relationships between the nodes is something that doesn't, at least in my case, come naturally. 

Granted, your approach leads to a much nicer implementation, but I think that it would take some time for a new user to develop the proper intuition on selection and unselection using a `Zipper` (specifically, when and how to use the short circuiting selection method).

Regarding "spooky action at a distance", although not nice, it takes the problem back to the user, and gives him/her the option to resolve it. On the other hand, merge strategies and the like, suffer from an inherent loss of information, which the user doesn't get to fully control.

To sum up, it's not that I think that we should, or even be able to, implement a mutable-like tree, it just that this notion is my source of inspiration when it comes the defining the proper behavior of a `Zipper`.

And as you said, we should wait for feedback and see whether we're on the right path...

**Comment by josharnold52** [comment]
> > Obviously intuitions change between different people, but I still think that when selecting on a tree, most people would (maybe naively) think of something along the lines of: "I want to select all elements with the name XXX and add the attribute YYY". When going through this process the thought of lineal relationships between the nodes is something that doesn't, at least in my case, come naturally.

I Agree.  That's the second of the "simple cases" from my earlier post.  Lineal conflicts don't cause any harm because you aren't touching the children of the nodes you select.   Ideally, you should be able to make this type of modification without worrying about conflicts, merge strategies, etc. 

The hard cases are when there are lineal conflicts _and_ you are manipulating the children of the nodes you select.  A very basic example would be "I want to select all elements named XXX and append an element YYY to their children", where XXX elements may share a lineal relationship.  Such cases really do require a merge strategy.  But I also think they admit other solutions, such as recursive application of short-circuit select, that are simpler and less error prone.  In essence, such solutions let you resolve the conflicts as you make them, rather than doing it all on the back end in a merge strategy,

Basically, I think the "sweet spot" for Zipper is the two use cases from my earlier post, which are "intuitively" conflict-free, and beyond that, it become a less useful tool due to its increased complexity.

**Comment by ncreep** [comment]
> ... admit other solutions, such as recursive application of short-circuit select

And that's exactly the sort of things our users will have to learn, I'm not even sure that at the moment I quite understand what that means.

Time will tell about the degree of sweetness of the spot we are at, let's hope you're right.

Next stop for me: trying to incorporate some more advanced selection methods, ideally with no changes to the `Zipper`'s implementation.

**Comment by josharnold52** [comment]
> > And that's exactly the sort of things our users will have to learn, I'm not even sure that at the moment I quite understand what that means.

```
/** Appends `a` to the children of every node in `xml` matching `sel` */
def appendChild(xml: Elem, sel: Selector[Elem], a: Elem):Elem = {
  val mods = (xml \\! sel) map { x =>
    val x2 = appendChild(x, sel, a)
    x2.copy(children = x2.children :+ a)
  }
  mods.unselect.head.asInstanceOf[Elem] //the asInstanceOf is annoying
}
```

That was a little harder to get right than I expected.  The fact that `\\!` skips the top level of nodes made it hard for me to keep my levels straight :-)  It sort of makes me want a deep-select that includes the top level...

Anyway, I hope the idea came through.  The equivalent function using `\\` requires a custom merge strategy, so there's a learning curve either way.  

> > Time will tell about the degree of sweetness of the spot we are at, let's hope you're right.

Well, I think we're in a good position regardless.   My modifications didn't remove any functionality, and we can always enhance.  

> > Next stop for me: trying to incorporate some more advanced selection methods, ideally with no changes to the Zipper's implementation.

Note that `PathCreator` has pretty much the same structure that it always did;  Only the datatypes changed.   `Zipper` itself should be completely general in terms of paths (aside from the issue of duplicates, which it never supported).   Of course there are plenty of reasons to make changes to `Zipper`, but I don't think you'll be forced into it due to a limitation on paths.

**Comment by ncreep** [comment]
Thanks for the quick illustration.
